apps/src/lib/node/ledger/shell/process_proposal.rs

```diff
@@ -293,12 +383,217 @@ mod test_process_proposal {
+        shell.storage.last_height = LAST_HEIGHT;
+        let (protocol_key, _) = wallet::defaults::validator_keys();
+        let vote_extension_digest = {
+            let validator_addr = wallet::defaults::validator_address();
+            let signed_vote_extension = {
+                let ext =
+                    VoteExtension::empty(LAST_HEIGHT, validator_addr.clone())
+                        .sign(&protocol_key);
+                assert!(ext.verify(&protocol_key.ref_to()).is_ok());
+                ext
+            };
+            // vote extension digest with no observed events
+            VoteExtensionDigest {
+                signatures: {
+                    let mut s = HashMap::new();
+                    s.insert(validator_addr, signed_vote_extension.sig);
+                    s
+                },
+                events: vec![],
+            }
+        };
+        let tx = ProtocolTxType::EthereumEvents(vote_extension_digest)
+            .sign(&protocol_key)
+            .to_bytes();
+        #[allow(clippy::redundant_clone)]
+        let request = ProcessProposal {
+            txs: vec![tx.clone(), tx],
+        };
+        let results = shell.process_proposal(request);
+        assert!(
+            matches!(results, Err(TestError::RejectProposal(s)) if s.len() == 2)
+        );
+    }
+
+    fn check_rejected_digest(
+        shell: &mut TestShell,
+        vote_extension_digest: VoteExtensionDigest,
+        protocol_key: common::SecretKey,
+    ) {
+        let tx = ProtocolTxType::EthereumEvents(vote_extension_digest)
+            .sign(&protocol_key)
+            .to_bytes();
+        let request = ProcessProposal { txs: vec![tx] };
+        let response = if let Err(TestError::RejectProposal(resp)) =
+            shell.process_proposal(request)
+        {
+            if let [resp] = resp.as_slice() {
+                resp.clone()
+            } else {
+                panic!("Test failed")
+            }
+        } else {
+            panic!("Test failed")
+        };
+        assert_eq!(
+            response.result.code,
+            u32::from(ErrorCodes::InvalidVoteExtension)
+        );
+    }
+
+    /// Test that if a proposal contains vote extensions with
+    /// invalid validator signatures, we reject it.
+    #[test]
+    fn test_drop_vext_digest_with_invalid_sigs() {
+        const LAST_HEIGHT: BlockHeight = BlockHeight(2);
+        let (mut shell, _, _) = test_utils::setup();
+        shell.storage.last_height = LAST_HEIGHT;
+        let (protocol_key, _) = wallet::defaults::validator_keys();
+        let vote_extension_digest = {
+            let addr = wallet::defaults::validator_address();
+            let event = EthereumEvent::TransfersToNamada {
+                nonce: 1u64.into(),
+                transfers: vec![],
+            };
+            let ext = {
+                // create a fake signature
+                let sig = common::Signature::Ed25519(ed25519::Signature(
```

Comment:
in `test_utils`, there is a function for generating random keypairs

Reply:
Boy I sure don't want my unit test to fail with a `1 / 2^256` chance